src/mem.rs

```diff
@@ -58,3 +62,110 @@ pub unsafe extern "C" fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i32 {
     }
     0
 }
+
+unsafe fn memcpy_element_unordered_atomic<T: Copy>(dest: *mut T, src: *const T, bytes: usize) {
+    let n = bytes / mem::size_of::<T>();
+    let mut i = 0;
+    while i < n {
+        atomic_store_unordered(dest.add(i), atomic_load_unordered(src.add(i)));
+        i += 1;
+    }
+}
+
+unsafe fn memmove_element_unordered_atomic<T: Copy>(dest: *mut T, src: *const T, bytes: usize) {
+    let n = bytes / mem::size_of::<T>();
+    if src < dest as *const T {
+        // copy from end
+        let mut i = n;
+        while i != 0 {
+            i -= 1;
+            atomic_store_unordered(dest.add(i), atomic_load_unordered(src.add(i)));
+        }
+    } else {
+        // copy from beginning
+        let mut i = 0;
+        while i < n {
+            atomic_store_unordered(dest.add(i), atomic_load_unordered(src.add(i)));
+            i += 1;
+        }
+    }
+}
+
+unsafe fn memset_element_unordered_atomic<T>(s: *mut T, c: u8, bytes: usize)
+where
+    T: Copy + From<u8> + Shl<u32, Output = T> + BitOr<T, Output = T>,
+{
+    let n = bytes / mem::size_of::<T>();
+    let mut x = T::from(c);
+    for _ in 1..mem::size_of::<T>() {
+        x = x << 8 | T::from(c);
+    }
+    let mut i = 0;
+    while i < n {
+        atomic_store_unordered(s.add(i), x);
+        i += 1;
+    }
+}
+
+macro_rules! trampoline {
+    ($(fn $fun:ident($($iid:ident : $ity:ty),+) {
+        $($body:tt)*
+    })+) => {
+        $(
+            #[cfg_attr(not(feature = "mangled-names"), no_mangle)]
+            pub unsafe extern "C" fn $fun($($iid: $ity),+) {
+                $($body)*
+            }
+        )+
+    }
+}
+
+trampoline! {
```

Comment:
Could this ideally try to use the existing `intrinsics!` macro?

Reply:
I was reluctant to do that because `intrinsics!` makes safe functions, while the raw pointer access means they very much are not.

Reply:
Ah I think that's ok, basically everything in this library is `unsafe` anyway so I think it's fine for the functions to look safe and they just have an initial `unsafe` block to call the right intrinsic.

Reply:
(also I'd prefer to prioritize reducing duplication around mangling and such)

Reply:
If you're sure. Change applied.